### PR TITLE
Enable zip64 to allow the creation of zip files larger than 4.3Gb

### LINKED ISSUE #579: Can not create zip files larger than 4.3Gb
### Steps to reproduce
1. Add files using `- (BOOL)writeData:(NSData *)data filename:(nullable NSString *)filename compressionLevel:(int)compressionLevel password:(nullable NSString *)password AES:(BOOL)aes;` until the archive reaches an expected size of around 5Gb
2. Check the content and size of the archive
3. The archive will be missing some files and the size is 4.3Gb instead of 5Gb

### Expected behavior
All files should be present in the archive, an error should appear, or the behaviour should be documented.

### Actual behavior
Files are missing from the zip archive without any notification of the user.

### Version of ZipArchive
v2.2.3

### Environmental information
Xcode 11.5 and iOS 13.5

What is the expected behaviour here?

### PR DESCRIPTION
**Pull request by joelfischerr**
Fixes #579 and #521.

It looks like the only thing that has to be changed is the flag indiciating whether zip64 should be used or not. Unpacking zip64 files seems to already be supported.

Should zip64 be enabled by default or should it be an option? I strongly favour making it the default choice since almost everything seems to support zip64.

I have created two tests to check that file sizes larger than 4.3Gb are supported. They fail when the zip64 flag is set to 0. 

I don't know whether I have added the tests correctly and they are not yet integrated with travis. Please let me know if any changes are required.

### CODE REVIEW

**Comment by Coeur** [comment]
i don't think it's supported by macOS default unarchiver. Consequently it should probably not be by default, and it needs to be an explicit decision by the developer or the user to generate non-standard archives.
I could be wrong... so be sure to test unzipping such large file by double clicking it on macOS.

File: ZipArchiveTests/ZipArchiveTests.m [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#import <XCTest/XCTest.h>
+#import "SSZipArchive.h"
+
+@interface SSZipArchiveTests : XCTestCase
+
+@end
+
+@implementation SSZipArchiveTests
+
+int twentyMb = 20971520;
```
Comment (Coeur) [inline ZipArchiveTests/ZipArchiveTests.m:18]: prefer `20 * 1024 * 1024` (or `20 * 0x100000`)

**Comment by joelfischerr** [comment]
> i don't think it's supported by macOS default unarchiver.

On my machine running macOS 10.15.5 (19F101) it works fine with the 6.29GB large file created by my test. But AES has to be disabled since that is not supported.

I'll try to take a look at the failing test tonight and will move my tests into the test suite of the Example.

File: ZipArchiveTests/ZipArchiveTests.m (hunk 1 above)
Comment (joelfischerr) [reply ZipArchiveTests/ZipArchiveTests.m]: ✔

**Comment by willson-chen** [comment]
> i don't think it's supported by macOS default unarchiver. Consequently it should probably not be by default, and it needs to be an explicit decision by the developer or the user to generate non-standard archives.
> I could be wrong... so be sure to test unzipping such large file by double clicking it on macOS.

Hi @Coeur , could you please offer some advice about if the zip64 should be supported by default? If not, then I think we need some more APIs such as `writeFoulderAtPath64` and `wireteDate64` to support zip64.

**Comment by willson-chen** [comment]
Hi @joelfischerr. Is this PR ready to be reviewed?

**Comment by joelfischerr** [comment]
@willson-chen Yes, it is ready to be reviewed.

Things to check before it is merged:
- [ ] Should zip64 be enabled by default.
- [ ] The new tests run longer than the old ones as a large zip file has to be created. Should they run on every PR?

**Comment by willson-chen** [comment]
> Things to check before it is merged:
> 
> * [ ]  Should zip64 be enabled by default.
> * [ ]  The new tests run longer than the old ones as a large zip file has to be created. Should they run on every PR?

Hi @Coeur @jhudsonWA . Could you please offer some advice about these two iteams?

**Comment by joelfischerr** [comment]
You can check that Archive Utility can unpack the zip64 files by running the tests in branch [expose-test-files](https://github.com/joelfischerr/ZipArchive/tree/expose-test-files) and then checking the file at the path that is printed during teardown.

**Comment by Coeur** [comment]
> > Things to check before it is merged:
> > 
> > * [ ]   Should zip64 be enabled by default.
> > * [ ]   The new tests run longer than the old ones as a large zip file has to be created. Should they run on every PR?
> 
> Hi @Coeur @jhudsonWA . Could you please offer some advice about these two items?

Well:
- Do the test to confirm wether zip64 is supported natively.
- Check the git history to see why I didn't implement it by default myself.
- Yes, tests should run on every PR. As far as they take less than 1 hour total, that's fine. If some single test is taking more than 5 minutes, then disable it for CI until one can find a faster way to do it (like by generating an almost zero-filled zip64 of some sort)

**Comment by joelfischerr** [comment]
I have changed the test to not run with different compression levels since that shouldn't make a difference. That should decrease the runtime of that test to less than 5 minutes.

@Coeur I requested the review from you because you did the first review. Are you the only one with the power to merge?
I'm currently fairly busy myself, if this had not been a one line change plus the unit tests I would have had to delay this until late this year. But I'd be happy to join later this year.

**Comment by willson-chen** [comment]
Hi @joelfischerr. I also have the power to merge. :)

I will spend some time to check the zip64 support. Thanks for your PR, and welcome to join the maintainers' team in the future.

**Comment by joelfischerr** [comment]
@willson-chen Any updates on the progress?

**Comment by willson-chen** [comment]
> @willson-chen Any updates on the progress?

Hi @joelfischerr. Sorry for my lateness. I am a little busy these days and will do the zip64 test before this weekend.

**Comment by willson-chen** [comment]
@joelfischerr I have reviewed your code and found that your two test cases can be merged into one, and at the same time change 300*20M to 240*20M, all above could save half of the test time.

```
diff --git a/Example/ObjectiveCExampleTests/SSZipArchiveTests.m b/Example/ObjectiveCExampleTests/SSZipArchiveTests.m
index 32e20a8..db038e9 100644
--- a/Example/ObjectiveCExampleTests/SSZipArchiveTests.m
+++ b/Example/ObjectiveCExampleTests/SSZipArchiveTests.m
@@ -608,22 +608,24 @@ int twentyMB = 20 * 1024 * 1024;
     }
 }
 
-// This tests whether the payload size of the zip file containing 6Gb of files with compression 0 is correct.
-- (void)testPayloadSizeCorrect {
-    long long int iterations = 300;
-    
+// This tests whether the payload size of the zip file containing 4.8Gb of files with compression 0 is correct,
+// and creates a zip file containing 240 20Mb sized files with multiple compression levels. It then unpacks the file and checks whether
+// the same number of files is present in the unpacked folder.
+- (void)testPayloadSizeCorrectAndUzipFileNo {
+    long long int iterations = 240;
+    NSString *unpackPath = [self _cachesPath:@"Unpacked/testFile"];
     NSNumber *goldenSize = [NSNumber numberWithLongLong:iterations * twentyMB];
     NSData *data = [self get20MbNSData];
     NSString *filenName = @"TestFile.zip";
     NSString *filePath = [NSString stringWithFormat:@"%@%@", [self _cachesPath:@""], filenName];
-    
+    NSString *password = @"TestPW";
     
     SSZipArchive *archive = [[SSZipArchive alloc] initWithPath:filePath];
     [archive open];
     
     for (int i = 0; i < iterations; i++) {
         NSString *fileName = [NSString stringWithFormat:@"File_%i", i];
-        [archive writeData:data filename:fileName compressionLevel:0 password:@"TestPW" AES:true];
+        [archive writeData:data filename:fileName compressionLevel:0 password:password AES:true];
     }
     
     bool close = [archive close];
@@ -633,34 +635,9 @@ int twentyMB = 20 * 1024 * 1024;
     XCTAssertTrue(close, "Should be able to close the archive.");
     XCTAssertTrue(fileSize.longLongValue == goldenSize.longLongValue,
                   "Payload size should be equal to the sum of the size of all files included.");
-}
-
-// This creates a zip file containing 300 20Mb sized files with multiple compression levels. It then unpacks the file and checks whether
-// the same number of files is present in the unpacked folder.
-- (void)testAllFilesPresent {
-    int iterations = 300;
-    NSString *unpackPath = [self _cachesPath:@"Unpacked/testFile"];
-    NSString *filenName = @"TestFile.zip";
-    NSString *filePath = [NSString stringWithFormat:@"%@%@", [self _cachesPath:@""], filenName];
-    NSString *password = @"TestPw";
-
-    NSData *data = [self get20MbNSData];
-
-    SSZipArchive *archive = [[SSZipArchive alloc] initWithPath:filePath];
-    [archive open];
-
-    for (int i = 0; i < iterations; i++) {
-        NSString *fileName = [NSString stringWithFormat:@"File_%i", i];
-        [archive writeData:data filename:fileName compressionLevel:0.0 password:password AES:true];
-    }
-
-    bool close = [archive close];
-
+    
     [SSZipArchive unzipFileAtPath:filePath toDestination:unpackPath overwrite:true password:password error:nil];
-
     long int noFiles = [[NSFileManager defaultManager] contentsOfDirectoryAtPath:unpackPath error:nil].count;
-
-    XCTAssertTrue(close, "Should be able to close the archive.");
     XCTAssertTrue(iterations == noFiles, "All files should be present in the exported directory");
 }
```

**Comment by joelfischerr** [comment]
@willson-chen Thanks for the review! Unfortunately applying your patch fails. But I have now allowed edits by maintainers, so you can push it directly if you want to.

**Comment by joelfischerr** [comment]
@willson-chen The builds seem to have been failing due to an incompatibility in cocoapods. I think I've been able to fix it.

**Comment by willson-chen** [comment]
@joelfischerr PR merged, thanks for your contribution.

**Pull request merged**